store/nosqlstore/driver.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package nosqlstore
+
+import (
+	"errors"
+)
+
+var ErrAlreadyExists = errors.New("already exists")
+var ErrNotFound = errors.New("not found")
+
+type RangeOrder int
+
+const (
+	RANGE_ORDER_ASC RangeOrder = iota
+	RANGE_ORDER_DESC
+)
+
+type RangeKey struct {
+	Min        []byte
+	ExcludeMin bool
+	Max        []byte
+	ExcludeMax bool
+	Order      RangeOrder
+	Offset     int
+	Limit      int
+}
+
+func Range(min, max []byte) *RangeKey {
+	return &RangeKey{Min: min, Max: max}
+}
+
+func RangeLessThan(key []byte) *RangeKey {
+	return &RangeKey{Max: key, ExcludeMax: true}
+}
+
+func RangeLessThanOrEqualTo(key []byte) *RangeKey {
+	return &RangeKey{Max: key}
+}
+
+func RangeEqualTo(key []byte) *RangeKey {
+	return &RangeKey{Min: key, Max: key}
+}
+
+func RangeGreaterThanOrEqual(key []byte) *RangeKey {
+	return &RangeKey{Min: key}
+}
+
+func RangeGreaterThan(key []byte) *RangeKey {
+	return &RangeKey{Min: key, ExcludeMin: true}
+}
+
+func RangeSubset(offset, limit int) *RangeKey {
+	return &RangeKey{
+		Offset: offset,
+		Limit:  limit,
+	}
+}
+
+func (k RangeKey) Subset(offset, limit int) *RangeKey {
+	k.Offset = offset
+	k.Limit = limit
+	return &k
+}
+
+type LookupResult struct {
+	TotalCount *int
+}
+
+type ObjectStore interface {
+	// Count returns the number of objects indexed by the given keys.
+	Count(index string, hashKey []byte, rangeKey *RangeKey) (int, error)
+
+	// Delete deletes an object by id.
+	Delete(id string) error
+
+	// Get gets an object by id.
+	Get(id string, dest interface{}) error
+
+	// Lookup gets the objects indexed by the given keys. If no range key is given, all results in
+	// ascending order from the beginning will be returned.
+	Lookup(index string, hashKey []byte, rangeKey *RangeKey, dest interface{}) error
+
+	// Upsert inserts or updates an object.
+	Upsert(id string, object interface{}) error
+
+	// AddIndex creates a named index for each object.
+	//
+	// The function should return two elements: The hash key can be used to perform lookups based on
+	// exact matches. The range key is used to order elements matches by the hash key. Objects with
+	// the same hash key cannot have the same range key.
+	//
+	// If the index does not already exist, re-indexing will occur.
+	AddIndex(name string, keys func(interface{}) (hashKey []byte, rangeKey []byte)) error
+}
```

Comment:
If you're familiar with DynamoDB, this hash key + range key indexing should feel right at home.

Reply:
I've read up on DynamoDB, and better appreciate the distinction being made here. Can you elaborate on how `AddIndex` would work in that context? It seems like you'd want to provide the attribute name to index, but it seems like this interface is geared toward providing the attribute value, as if the index is being create item by item. Am I misunderstanding how the interface will be used?

Reply:
A single attribute name wouldn't be enough to describe the thing being indexed. Often you want more complex indexes (e.g. `Encode(wh.DeleteAt, wh.Id)`). It's probably also preferable for implementations of this method to not have to resort to `reflect`. So this is what I went with. As far as I can tell you're not misunderstanding anything.

Reply:
Alas, I think I am still confused! Thank you for your patience :)

In a world where we decide to support MongoDB, we'd be implementing a driver to invoke https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/ to create the indices. What would a pseudo-code implementation of that look like?

I can see how this method allows me to build an in-memory index for each item's value, but it's not clear how I can take the parameters passed to this method and achieve the same thing where someone else is building the index.

Reply:
Document stores such as MongoDB would probably just add a new field to the documents that's assigned the return value of the `keys` function here (and the index would be created on that).

Reply:
Maybe it would clear things up if I just went ahead and created driver implementations for other database types?

Reply:
Yeah, ^ is more like what I had imagined.